File: Python/Structures/Deque.py
class Node:
    def __init__(self, data: int) -> None:
        self.data = data
        self.next: Node = None
        self.prev: Node = None


class Deque:
    def __init__(self) -> None:
        self.size: int = 0
        self.front: Node = None
        self.back: Node = None

    def push_back(self, data: int) -> None:
        node = Node(data)
        if self.is_empty():
            self.front = node
            self.back = node
        else:
            self.back.next = node
            node.prev = self.back
            self.back = node

        self.size += 1

    def push_front(self, data: int) -> None:
        node = Node(data)
        if self.is_empty():
            self.front = node
            self.back = node
        else:
            self.front.next = node
            node.prev = self.front
            self.front = node

        self.size += 1

    def pop_back(self) -> None:
        ptr = self.back.prev
        self.back = ptr
        self.size -= 1

    def pop_front(self) -> None:
        ptr = self.front.prev
        self.front = ptr
        self.size -= 1

    def display(self) -> None:
        ptr = self.front
        # print(self.front.next is None)
        while ptr.prev != None:
            print(ptr.data)
            ptr = ptr.prev

    def display_reverse(self) -> None:
        ptr = self.back
        # print(self.back.next is None)
        while ptr.next != None:
            print(ptr.data)
            ptr = ptr.next

    def is_empty(self) -> bool:
        return self.front == None and self.size == 0


if __name__ == '__main__':
    dq = Deque()
    for i in range(0, 9):
        dq.push_front(i)
    # dq.push_back(25) problematic
    dq.pop_front()
    # dq.pop_back()

    dq.display()
    # dq.display_reverse() problematic

File: Dart/Structures/Deque.dart
class Node {
  Node? next;
  Node? prev;
  late int data;
  Node(int data) {
    this.data = data;
    this.prev = null;
    this.next = null;
  }
}

class Deque {
  Node? front;
  Node? back;
  late int size;
  Deque() {
    this.size = 0;
    this.front = null;
    this.back = null;
  }

  void pushBack(int data) {
    Node node = new Node(data);
    if (isEmpty()) {
      this.front = node;
      this.back = node;
    } else {
      this.back?.next = node;
      node.prev = this.back;
      this.back = node;
    }
    ++this.size;
  }

  void pushFront(int data) {
    Node node = new Node(data);
    if (isEmpty()) {
      this.front = node;
      this.back = node;
    } else {
      this.front?.next = node;
      node.prev = this.front;
      this.front = node;
    }
    ++this.size;
  }

  void popBack() {
    // insert error handler if empty
    this.back = this.back?.prev;
    --this.size;
  }

  void popFront() {
    // insert error handler if empty
    this.front = this.front?.prev;
    --this.size;
  }

  void display() {
    Node? ptr = this.front;
    while (ptr?.next != null) {
      print("${ptr?.data}");
      ptr = ptr?.next;
    }
  }

  void displayReverse() {}

  bool isEmpty() {
    return this.front == null && this.size == 0;
  }

  int sizeList() {
    return this.size;
  }
}
